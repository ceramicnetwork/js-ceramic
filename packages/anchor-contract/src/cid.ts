import type { EventFragment } from '@ethersproject/abi'
import type { Log, TransactionResponse } from '@ethersproject/providers'
import { create as createMultihash } from 'multiformats/hashes/digest'
import { CID } from 'multiformats/cid'
import { fromString } from 'uint8arrays'

import { contractInterface } from './interface.js'

const SHA256_CODE = 0x12
const DAG_CBOR_CODE = 0x71
const V1_PROOF_TYPE = 'f(bytes32)'

export function createCidFromHexValue(value: string): CID {
  const multihash = createMultihash(SHA256_CODE, fromString(value.slice(2), 'base16'))
  return CID.create(1, DAG_CBOR_CODE, multihash)
}

export function getCidFromAnchorEventLog(log: Log): CID {
  const decodedArgs = contractInterface.decodeEventLog(
    contractInterface.events['DidAnchor'] as EventFragment,
    log.data
  )
  return createCidFromHexValue(decodedArgs[1])
}

const getCidFromV0Transaction = (txResponse: TransactionResponse): CID => {
  const withoutPrefix = txResponse.data.replace(/^(0x0?)/, '')
  return CID.decode(fromString(withoutPrefix.slice(1), 'base16'))
}

const getCidFromV1Transaction = (txResponse: TransactionResponse): CID => {
  const decodedArgs = contractInterface.decodeFunctionData('anchorDagCbor', txResponse.data)
  return createCidFromHexValue(decodedArgs[0])
}

/**
 * Parses the transaction data to recover the CID.
 * @param txType transaction type of the anchor proof. Currently support `raw` and `f(bytes32)`
 * @param txResponse the retrieved transaction from the ethereum blockchain
 * @returns
 */
export const getCidFromTransaction = (txType: string, txResponse: TransactionResponse): CID => {
  if (txType === V1_PROOF_TYPE) {
    return getCidFromV1Transaction(txResponse)
  } else {
    return getCidFromV0Transaction(txResponse)
  }
}
