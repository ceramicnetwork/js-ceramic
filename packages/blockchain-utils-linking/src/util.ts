export interface LinkProof {
  version: number;
  message: string;
  signature: string;
  account: string;
  did?: string;
  timestamp?: number;
  address?: string;
  type?: string;
  chainId?: number;
}

export interface RpcMessage {
  jsonrpc: string;
  id: number;
  method: string;
  params: any;
}

export interface ConsentMessage {
  message: string;
  timestamp?: number;
}

export function getConsentMessage(
  did: string,
  addTimestamp = true
): ConsentMessage {
  const res: any = {
    message: "Create a new account link to your identity." + "\n\n" + did,
  };
  if (addTimestamp) {
    res.timestamp = Math.floor(Date.now() / 1000);
    res.message += " \n" + "Timestamp: " + res.timestamp;
  }
  return res;
}

export function encodeRpcMessage(method: string, params?: any): any {
  return {
    jsonrpc: "2.0",
    id: 1,
    method,
    params,
  };
}
