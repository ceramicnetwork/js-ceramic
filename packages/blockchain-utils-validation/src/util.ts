import { AccountId } from 'caip'

export function normalizeAccountId(accountId: string | AccountId): AccountId {
  // Handle legacy CAIP-10 representation
  if (typeof accountId === 'string' && accountId.includes('@')) {
    const [address, chainId] = accountId.split('@')
    if (!address || !chainId) {
      throw new Error(`Invalid accountId provided`)
    }

    return new AccountId({ address, chainId })
  }
  return new AccountId(accountId)
}
