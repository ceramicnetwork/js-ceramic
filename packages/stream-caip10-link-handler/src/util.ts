// NOTE: This differs from `normalizeAccountId` in `stream-caip10-link`
// This converts from CAIP-10 repr to the legacy repr
// The former converts from legacy to the current repr
export function normalizeAccountId(accountId: string): string {
  if (!accountId.includes('@')) {
    const _accountSplit = accountId.split(':')
    const address = _accountSplit.pop()
    const chainId = _accountSplit.join(':')
    return `${address}@${chainId}`
  }
  return accountId
}
