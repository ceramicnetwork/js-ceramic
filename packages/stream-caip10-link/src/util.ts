import { AccountId } from 'caip'

export function normalizeAccountId(accountId: string | AccountId): AccountId {
  if (typeof accountId === 'string' && accountId.includes('@')) {
    const [address, chainId] = accountId.split('@')
    if (!address || !chainId) {
      throw new Error(`Invalid accountId provided`)
    }

    return new AccountId({ address, chainId })
  }
  return new AccountId(accountId)
}
