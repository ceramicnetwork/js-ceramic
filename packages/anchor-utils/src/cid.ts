import type { EventFragment } from '@ethersproject/abi'
import type { Log, TransactionResponse } from '@ethersproject/providers'
import { create as createMultihash } from 'multiformats/hashes/digest'
import { CID } from 'multiformats/cid'
import { fromString } from 'uint8arrays'

import { contractInterface } from './interface.js'

export const SHA256_CODE = 0x12
export const KECCAK_256_CODE = 0x1b
export const DAG_CBOR_CODE = 0x71
export const ETH_TX_CODE = 0x93

export function createCidFromHexValue(hexValue: string): CID {
  const multihash = createMultihash(SHA256_CODE, fromString(hexValue.slice(2), 'base16'))
  return CID.create(1, DAG_CBOR_CODE, multihash)
}

export function getCidFromAnchorEventLog(log: Log): CID {
  const decodedArgs = contractInterface.decodeEventLog(
    contractInterface.events['DidAnchor(address,bytes32)'] as EventFragment,
    log.data
  )
  return createCidFromHexValue(decodedArgs[1])
}

const getCidFromV0Transaction = (txResponse: TransactionResponse): CID => {
  const withoutPrefix = txResponse.data.replace(/^(0x0?)/, '')
  return CID.decode(fromString(withoutPrefix.slice(1), 'base16'))
}

const getCidFromV1Transaction = (txResponse: TransactionResponse): CID => {
  const decodedArgs = contractInterface.decodeFunctionData('anchorDagCbor', txResponse.data)
  return createCidFromHexValue(decodedArgs[0])
}

const V1_PROOF_TYPE = 'f(bytes32)'

/**
 * Parses the transaction data to recover the CID.
 * @param txType transaction type of the anchor proof. Currently support `raw` and `f(bytes32)`
 * @param txResponse the retrieved transaction from the ethereum blockchain
 * @returns
 */
export const getCidFromTransaction = (txType: string, txResponse: TransactionResponse): CID => {
  if (txType === V1_PROOF_TYPE) {
    return getCidFromV1Transaction(txResponse)
  } else {
    return getCidFromV0Transaction(txResponse)
  }
}

/**
 * Converts ETH address to CID
 * @param hash - ETH hash
 */
export function convertEthHashToCid(hexValue: string): CID {
  const multihash = createMultihash(KECCAK_256_CODE, fromString(hexValue.slice(2), 'base16'))
  return CID.create(1, ETH_TX_CODE, multihash)
}
