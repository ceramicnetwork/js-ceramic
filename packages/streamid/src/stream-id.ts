import { CID } from 'multiformats/cid'
import * as Block from 'multiformats/block'
import { base36 } from 'multiformats/bases/base36'
import { sha256 as hasher } from 'multiformats/hashes/sha2'
import varint from 'varint'
import * as codec from '@ipld/dag-cbor'
import { concat as uint8ArrayConcat } from 'uint8arrays'
import { STREAMID_CODEC } from './constants.js'
import { Memoize } from 'typescript-memoize'
import type { StreamRef } from './stream-ref.js'
import { StreamType } from './stream-type.js'
import { tryCatch } from './try-catch.util.js'
import * as parsing from './stream-ref-parsing.js'

export class InvalidStreamIDBytesError extends Error {
  constructor(bytes: Uint8Array) {
    super(`Invalid StreamID bytes ${base36.encode(bytes)}: contains commit`)
  }
}

export class InvalidStreamIDStringError extends Error {
  constructor(input: string) {
    super(`Invalid StreamID string ${input}: contains commit`)
  }
}

/**
 * Parse StreamID from bytes representation.
 *
 * @param bytes - bytes representation of StreamID.
 * @throws error on invalid input
 * @see StreamID#bytes
 */
function fromBytes(bytes: Uint8Array): StreamID {
  const parsed = parsing.fromBytes(bytes, 'StreamID')
  if (parsed.kind === 'stream-id') {
    return new StreamID(parsed.type, parsed.genesis)
  }
  throw new InvalidStreamIDBytesError(bytes)
}

/**
 * Same as fromBytes, but returns an Error instance rather than throwing if there is a problem
 * with the input.
 * Note that some exceptions can still be thrown in certain cases, if they come from lower-level
 * libraries like multibase, for example.
 * @param bytes
 */
function fromBytesNoThrow(bytes: Uint8Array): StreamID | Error {
  return tryCatch(() => fromBytes(bytes))
}

/**
 * Parse StreamID from string representation.
 *
 * @param input - string representation of StreamID, be it base36-encoded string or URL.
 * @see StreamID#toString
 * @see StreamID#toUrl
 */
function fromString(input: string): StreamID {
  const parsed = parsing.fromString(input, 'StreamID')
  if (parsed.kind === 'stream-id') {
    return new StreamID(parsed.type, parsed.genesis)
  }
  throw new InvalidStreamIDStringError(input)
}

/**
 * Same as fromString but returns an Error instance rather than throwing if there is a problem
 * with the input.
 * Note that some exceptions can still be thrown in certain cases, if they come from lower-level
 * libraries like multibase, for example.
 * @param input
 */
function fromStringNoThrow(input: string): StreamID | Error {
  return tryCatch(() => fromString(input))
}

const TAG = Symbol.for('@ceramicnetwork/streamid/StreamID')

/**
 * Stream identifier, no commit information included.
 *
 * Contains stream type and CID of genesis commit.
 *
 * Encoded as `<multibase-prefix><multicodec-streamid><type><genesis-cid-bytes>`.
 *
 * String representation is base36-encoding of the bytes above.
 */
export class StreamID implements StreamRef {
  protected readonly _tag = TAG

  private readonly _type: number
  private readonly _cid: CID

  static fromBytes = fromBytes
  static fromBytesNoThrow = fromBytesNoThrow
  static fromString = fromString
  static fromStringNoThrow = fromStringNoThrow

  // WORKARDOUND Weird replacement for Symbol.hasInstance due to
  // this old bug in Babel https://github.com/babel/babel/issues/4452
  // which is used by CRA, which is widely popular.
  static isInstance(instance: any): instance is StreamID {
    return typeof instance === 'object' && '_tag' in instance && instance._tag === TAG
  }

  /**
   * Create a new StreamID.
   *
   * @param {string|number}      type       the stream type
   * @param {CID|string}         cid
   *
   * @example
   * ```typescript
   * new StreamID('tile', 'bagcqcerakszw2vsovxznyp5gfnpdj4cqm2xiv76yd24wkjewhhykovorwo6a');
   * new StreamID('tile', cid);
   * new StreamID(0, cid);
   * ```
   */
  constructor(type: string | number, cid: CID | string) {
    if (!(type || type === 0)) throw new Error('StreamID constructor: type required')
    if (!cid) throw new Error('StreamID constructor: cid required')
    this._type = typeof type === 'string' ? StreamType.codeByName(type) : type
    this._cid = typeof cid === 'string' ? CID.parse(cid) : cid
  }

  /**
   * Create a streamId from a genesis commit.
   *
   * @param {string|number}         type       the stream type
   * @param {Record<string, any>}   genesis    a genesis commit
   *
   * @example
   * ```typescript
   * const streamId = StreamID.fromGenesis('tile', {
   *   header: {
   *     controllers:['did:3:kjz...'],
   *     family: 'IDX'
   *   }
   * });
   * ```
   */
  static async fromGenesis(type: string | number, genesis: Record<string, any>): Promise<StreamID> {
    const block = await Block.encode({ value: genesis, codec, hasher })
    return new StreamID(type, block.cid)
  }

  /**
   * Stream type code
   */
  get type(): number {
    return this._type
  }

  /**
   * Stream type name
   */
  @Memoize()
  get typeName(): string {
    return StreamType.nameByCode(this._type)
  }

  /**
   * Genesis commits CID
   */
  get cid(): CID {
    return this._cid
  }

  /**
   * Bytes representation of StreamID.
   */
  @Memoize()
  get bytes(): Uint8Array {
    const codec = varint.encode(STREAMID_CODEC)
    const type = varint.encode(this.type)

    return uint8ArrayConcat([codec, type, this.cid.bytes])
  }

  /**
   * Copy of self. Exists to maintain compatibility with CommitID.
   * @readonly
   */
  @Memoize()
  get baseID(): StreamID {
    return new StreamID(this._type, this._cid)
  }

  /**
   * Compare equality with another StreamID.
   */
  equals(other: StreamID): boolean {
    if (StreamID.isInstance(other)) {
      return this.type === other.type && this.cid.equals(other.cid)
    } else {
      return false
    }
  }

  /**
   * Encode the StreamID into a string.
   */
  @Memoize()
  toString(): string {
    return base36.encode(this.bytes)
  }

  /**
   * Encode the StreamID into a base36 url.
   */
  @Memoize()
  toUrl(): string {
    return `ceramic://${this.toString()}`
  }

  /**
   * StreamId(k3y52l7mkcvtg023bt9txegccxe1bah8os3naw5asin3baf3l3t54atn0cuy98yws)
   */
  [Symbol.for('nodejs.util.inspect.custom')](): string {
    return `StreamID(${this.toString()})`
  }

  /**
   * String representation of StreamID.
   */
  [Symbol.toPrimitive](): string {
    return this.toString()
  }
}
