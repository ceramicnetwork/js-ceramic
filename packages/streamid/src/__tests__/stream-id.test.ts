import { jest } from '@jest/globals'
import { StreamID } from '../stream-id.js'
import { CID } from 'multiformats/cid'
import * as util from 'util'
import { base36 } from 'multiformats/bases/base36'

const CID_STRING = 'bagcqcerakszw2vsovxznyp5gfnpdj4cqm2xiv76yd24wkjewhhykovorwo6a'
const STREAM_ID_STRING = 'kjzl6cwe1jw147dvq16zluojmraqvwdmbh61dx9e0c59i344lcrsgqfohexp60s'
const STREAM_ID_URL = 'ceramic://kjzl6cwe1jw147dvq16zluojmraqvwdmbh61dx9e0c59i344lcrsgqfohexp60s'
const STREAM_ID_BYTES = base36.decode(STREAM_ID_STRING)
const STREAM_ID_LEGACY = '/ceramic/kjzl6cwe1jw147dvq16zluojmraqvwdmbh61dx9e0c59i344lcrsgqfohexp60s'

const STREAM_ID_WITH_COMMIT =
  'k1dpgaqe3i64kjqcp801r3sn7ysi5i0k7nxvs7j351s7kewfzr3l7mdxnj7szwo4kr9mn2qki5nnj0cv836ythy1t1gya9s25cn1nexst3jxi5o3h6qprfyju'
const STREAM_ID_WITH_COMMIT_BYTES = base36.decode(STREAM_ID_WITH_COMMIT)
const STREAM_ID_WITH_0_COMMIT = 'k3y52l7qbv1frxwipl4hp7e6jlu4f6u8upm2xv0irmedfkm5cnutmezzi3u7mytj4'
const STREAM_ID_WITH_0_COMMIT_BYTES = base36.decode(STREAM_ID_WITH_0_COMMIT)
const STREAM_ID_WITH_COMMIT_LEGACY =
  '/ceramic/kjzl6cwe1jw147dvq16zluojmraqvwdmbh61dx9e0c59i344lcrsgqfohexp60s?commit=bagjqcgzaday6dzalvmy5ady2m5a5legq5zrbsnlxfc2bfxej532ds7htpova'
const STREAM_ID_WITH_0_COMMIT_LEGACY =
  '/ceramic/kjzl6cwe1jw147dvq16zluojmraqvwdmbh61dx9e0c59i344lcrsgqfohexp60s?commit=0'

describe('constructor', () => {
  test('create by parts (type:int, cid:cid)', () => {
    const type = 0
    const cid = CID.parse(CID_STRING)
    const streamid = new StreamID(type, cid)

    expect(streamid.type).toEqual(type)
    expect(streamid.cid.toString()).toEqual(CID_STRING)
    expect(streamid.toString()).toMatchSnapshot()
  })
})

describe('.fromBytes', () => {
  test('create from bytes ', () => {
    const streamid = StreamID.fromBytes(STREAM_ID_BYTES)

    expect(streamid.type).toEqual(0)
    expect(streamid.cid.toString()).toEqual(CID_STRING)
    expect(streamid.toString()).toMatchSnapshot()
  })

  test('fail to create from garbage bytes', () => {
    expect(() =>
      StreamID.fromBytes(
        CID.parse('bagcqcerakszw2vsovxznyp5gfnpdj4cqm2xiv76yd24wkjewhhykovorwo6a').bytes
      )
    ).toThrow()
  })

  test('create from bytes including commit: fail', () => {
    expect(() => StreamID.fromBytes(STREAM_ID_WITH_COMMIT_BYTES)).toThrow()
  })

  test('create from bytes including commit 0: fail', () => {
    expect(() => StreamID.fromBytes(STREAM_ID_WITH_0_COMMIT_BYTES)).toThrow()
  })

  test('roundtrip streamID bytes', () => {
    const streamid = new StreamID('tile', CID_STRING)
    const streamid2 = StreamID.fromBytes(streamid.bytes)
    expect(streamid.toString()).toEqual(streamid2.toString())
  })
})

describe('.fromString', () => {
  test('create from string', () => {
    const streamid = StreamID.fromString(STREAM_ID_STRING)

    expect(streamid.type).toEqual(0)
    expect(streamid.cid.toString()).toEqual(CID_STRING)
    expect(streamid.toString()).toMatchSnapshot()
  })

  test('create from string including commit: fail', () => {
    expect(() => StreamID.fromString(STREAM_ID_WITH_COMMIT)).toThrow()
  })

  test('create from string including commit 0: fail', () => {
    expect(() => StreamID.fromString(STREAM_ID_WITH_0_COMMIT)).toThrow()
  })

  test('create from url string', () => {
    const streamid = StreamID.fromString(STREAM_ID_URL)

    expect(streamid.type).toEqual(0)
    expect(streamid.cid.toString()).toEqual(CID_STRING)
    expect(streamid.toString()).toMatchSnapshot()
  })

  test('create from legacy string "/ceramic/"', () => {
    const streamid = StreamID.fromString(STREAM_ID_LEGACY)

    expect(streamid.type).toEqual(0)
    expect(streamid.cid.toString()).toEqual(CID_STRING)
    expect(streamid.toString()).toMatchSnapshot()
  })

  test('create from legacy string "/ceramic/" with commit param "?commit=": fail', () => {
    expect(() => StreamID.fromString(STREAM_ID_WITH_COMMIT_LEGACY)).toThrow()
  })

  test('create from legacy string "/ceramic/" with commit param "?commit=0"', () => {
    expect(() => StreamID.fromString(STREAM_ID_WITH_0_COMMIT_LEGACY)).toThrow()
  })

  test('roundtrip streamID string', () => {
    const streamid = new StreamID('tile', CID_STRING)
    const streamid2 = StreamID.fromString(streamid.toString())
    expect(streamid.toString()).toEqual(streamid2.toString())
  })
})

test('.bytes', () => {
  const streamid = new StreamID('tile', CID_STRING)
  const bytes = streamid.bytes
  expect(bytes).toBeDefined()
  expect(bytes instanceof Uint8Array).toEqual(true)
  expect(bytes).toMatchSnapshot()
})

test('.typeName if registered', () => {
  const streamid = new StreamID('tile', CID_STRING)
  expect(streamid.typeName).toEqual('tile')
  const streamid2 = new StreamID(10, CID_STRING)
  expect(() => streamid2.typeName).toThrowErrorMatchingSnapshot()
})

test('.toString()', () => {
  const streamid = new StreamID('tile', CID_STRING)
  const str = streamid.toString()
  expect(str).toBeDefined()
  expect(str).toMatchSnapshot()
})

test('.toUrl()', () => {
  const streamid = new StreamID('tile', CID_STRING)
  const str = streamid.toUrl()
  expect(str).toBeDefined()
  expect(str.includes('ceramic://')).toEqual(true)
  expect(str).toMatchSnapshot()
})

test('equals other StreamID', () => {
  const streamid = new StreamID('tile', CID_STRING)
  const streamid2 = new StreamID('tile', CID_STRING)
  const streamid3 = new StreamID('caip10-link', CID_STRING)
  expect(streamid.equals(streamid2)).toEqual(true)
  expect(streamid.equals(streamid3)).toEqual(false)
})

test('nodejs inspect', () => {
  const streamid = new StreamID('tile', CID_STRING)
  expect(util.inspect(streamid)).toMatchSnapshot()
})

test('to primitive', () => {
  const streamid = new StreamID('tile', CID_STRING)
  expect(`${streamid}`).toEqual(streamid.toString())
  expect(+streamid).toBeNaN()
  expect(streamid + '').toEqual(streamid.toString())
})

test('instanceof', () => {
  const streamId = StreamID.fromString(STREAM_ID_STRING)
  const instanceSpy = jest.spyOn(StreamID, 'isInstance')
  expect(StreamID.isInstance(streamId)).toBeTruthy()
  expect(instanceSpy).toBeCalledWith(streamId)
})
