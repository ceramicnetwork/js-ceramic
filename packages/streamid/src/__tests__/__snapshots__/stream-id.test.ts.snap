// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`.bytes 1`] = `
Uint8Array [
  206,
  1,
  0,
  1,
  133,
  1,
  18,
  32,
  84,
  179,
  109,
  86,
  78,
  173,
  242,
  220,
  63,
  166,
  43,
  94,
  52,
  240,
  80,
  102,
  174,
  138,
  255,
  216,
  30,
  185,
  101,
  36,
  150,
  57,
  240,
  167,
  85,
  209,
  179,
  188,
]
`;

exports[`.fromBytes create from bytes  1`] = `"kjzl6cwe1jw147dvq16zluojmraqvwdmbh61dx9e0c59i344lcrsgqfohexp60s"`;

exports[`.fromString create from legacy string "/ceramic/" 1`] = `"kjzl6cwe1jw147dvq16zluojmraqvwdmbh61dx9e0c59i344lcrsgqfohexp60s"`;

exports[`.fromString create from string 1`] = `"kjzl6cwe1jw147dvq16zluojmraqvwdmbh61dx9e0c59i344lcrsgqfohexp60s"`;

exports[`.fromString create from url string 1`] = `"kjzl6cwe1jw147dvq16zluojmraqvwdmbh61dx9e0c59i344lcrsgqfohexp60s"`;

exports[`.toString() 1`] = `"kjzl6cwe1jw147dvq16zluojmraqvwdmbh61dx9e0c59i344lcrsgqfohexp60s"`;

exports[`.toUrl() 1`] = `"ceramic://kjzl6cwe1jw147dvq16zluojmraqvwdmbh61dx9e0c59i344lcrsgqfohexp60s"`;

exports[`.typeName if registered 1`] = `"No stream type registered for index 10"`;

exports[`constructor create by parts (type:int, cid:cid) 1`] = `"kjzl6cwe1jw147dvq16zluojmraqvwdmbh61dx9e0c59i344lcrsgqfohexp60s"`;

exports[`nodejs inspect 1`] = `"StreamID(kjzl6cwe1jw147dvq16zluojmraqvwdmbh61dx9e0c59i344lcrsgqfohexp60s)"`;
